### PR TITLE
[ob1k-cache] added monitoring cache delegate (plus tests)

### PR DESCRIPTION
**Pull request by hunchback**
including:
- monitoring cache delegate
- tested with memcacheClient
- added unit tests

### CODE REVIEW

File: ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/MonitoringCacheDelegate.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+      total.inc();
+      if (result != null) {
+        hits.inc();
+      }
+      return result;
+    }
+  };
+
+  public MonitoringCacheDelegate(final TypedCache<K, V> delegate, String cacheName, final MetricFactory metricFactory) {
+    Preconditions.checkNotNull(delegate);
```
Comment (marenzo) [inline ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/MonitoringCacheDelegate.java]: Preconditions returns the tested object, so
this.delegate = Preconditions.checkNotNull(delegate);

File: ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/MonitoringCacheDelegate.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+      total.inc();
+      if (result != null) {
+        hits.inc();
+      }
+      return result;
+    }
+  };
+
+  public MonitoringCacheDelegate(final TypedCache<K, V> delegate, String cacheName, final MetricFactory metricFactory) {
+    this.delegate = Preconditions.checkNotNull(delegate);
```
Comment (marenzo) [inline ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/MonitoringCacheDelegate.java]: static import checkNotNull
add message ("delegate may not be null")

File: ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/MonitoringCacheDelegate.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    Preconditions.checkNotNull(metricFactory);
```
Comment (marenzo) [inline ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/MonitoringCacheDelegate.java]: move up

File: ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/MonitoringCacheDelegate.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    final String component = delegate.getClass().getSimpleName() + "." + cacheName;
+
+    getAsyncMetrics = new AsyncOperationMetrics<>(metricFactory, component + ".getAsync");
+    getBulkAsyncMetrics = new AsyncOperationMetrics<>(metricFactory, component + ".getBulkAsync");
+    setAsyncMetrics = new AsyncOperationMetrics<>(metricFactory, component + ".setAsync");
+    setBulkAsyncMetrics = new AsyncOperationMetrics<>(metricFactory, component + ".setBulkAsync");
+    deleteAsyncMetrics = new AsyncOperationMetrics<>(metricFactory, component + ".deleteAsync");
+
+    final String CACHE = ".Cache";
```
Comment (marenzo) [inline ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/MonitoringCacheDelegate.java]: final String componentCache = componant + ".cache"

File: ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/AsyncOperationMetrics.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+      @Override
+      public ComposableFuture<T> handle(final Try<T> result) {
+        tc.stop();
+        if (!result.isSuccess()) {
+          errors.inc();
+          if (result.getError() instanceof TimeoutException) {
+            timeouts.inc();
+          }
+        }
+        return ComposableFutures.fromTry(result);
```
Comment (marenzo) [inline ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/AsyncOperationMetrics.java]: static import

File: ob1k-cache/src/test/java/com/outbrain/ob1k/cache/metrics/MonitoringCacheDelegateTest.java [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+public class MonitoringCacheDelegateTest {
+  private static final MetricFactory metricFactory = mock(MetricFactory.class, withSettings().defaultAnswer(RETURNS_MOCKS));
+  private static final TypedCache<String, String> cache = mock(TypedCache.class);
+  private static final TypedCache<String, String> monitor = new MonitoringCacheDelegate<>(cache, MonitoringCacheDelegateTest.class.getSimpleName(), metricFactory);
+  private static final String SUCCESS_MSG = "failed success test";
+  private static final String ERROR_MSG = "failed error test";
+  private static final String TIMEOUT_MSG = "failed timeout test";
+
+  public static class getAsyncTest {
+    private final String key = "key1", value = "value1";
```
Comment (marenzo) [inline ob1k-cache/src/test/java/com/outbrain/ob1k/cache/metrics/MonitoringCacheDelegateTest.java]: two fields

File: ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/MonitoringCacheDelegate.java (hunk 2 above)
Comment (hunchback) [inline ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/MonitoringCacheDelegate.java]: corrected

File: ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/MonitoringCacheDelegate.java (hunk 3 above)
Comment (hunchback) [inline ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/MonitoringCacheDelegate.java]: corrected

File: ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/MonitoringCacheDelegate.java (hunk 4 above)
Comment (hunchback) [inline ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/MonitoringCacheDelegate.java]: corrected

File: ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/AsyncOperationMetrics.java (hunk 5 above)
Comment (hunchback) [inline ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/AsyncOperationMetrics.java]: done

File: ob1k-cache/src/test/java/com/outbrain/ob1k/cache/metrics/MonitoringCacheDelegateTest.java (hunk 6 above)
Comment (hunchback) [inline ob1k-cache/src/test/java/com/outbrain/ob1k/cache/metrics/MonitoringCacheDelegateTest.java]: done

**Pull request merged**